recipes/mp-units/2.2.0/conanfile.py

```diff
@@ -0,0 +1,267 @@
+        "units",
+        "dimensions",
+        "quantities",
+        "dimensional-analysis",
+        "physical-quantities",
+        "physical-units",
+        "system-of-units",
+        "system-of-quantities",
+        "isq",
+        "si",
+        "library",
+        "quantity-manipulation",
+    )
+    license = "MIT"
+    url = "https://github.com/conan-io/conan-center-index"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "cxx_modules": ["auto", True, False],
+        "std_format": ["auto", True, False],
+        "string_view_ret": ["auto", True, False],
+        "no_crtp": ["auto", True, False],
+        "contracts": ["none", "gsl-lite", "ms-gsl"],
+        "freestanding": [True, False],
+    }
+    default_options = {
+        "cxx_modules": "auto",
+        "std_format": "auto",
+        "string_view_ret": "auto",
+        "no_crtp": "auto",
+        "contracts": "gsl-lite",
+        "freestanding": False,
+    }
+    tool_requires = "cmake/[>=3.29]"
+    package_type = "header-library"
+    no_copy_source = True
+
+    @property
+    def _feature_compatibility(self):
+        return {
+            "minimum_support": {
+                "std": "20",
+                "compiler": {
+                    "gcc": "12",
+                    "clang": "16",
+                    "apple-clang": "15",
+                    "msvc": "",
+                },
+            },
+            "std_format": {
+                "std": "20",
+                "compiler": {
+                    "gcc": "13",
+                    "clang": "17",
+                    "apple-clang": "",
+                    "msvc": "",
+                },
+            },
+            "cxx_modules": {
+                "std": "20",
+                "compiler": {"gcc": "", "clang": "17", "apple-clang": "", "msvc": ""},
+            },
+            "static_constexpr_vars_in_constexpr_func": {
+                "std": "23",
+                "compiler": {"gcc": "13", "clang": "17", "apple-clang": "", "msvc": ""},
+            },
+            "explicit_this": {
+                "std": "23",
+                "compiler": {
+                    "gcc": "14",
+                    "clang": "18",
+                    "apple-clang": "",
+                    "msvc": "",
+                },
+            },
+        }
+
+    @property
+    def _option_feature_map(self):
+        return {
+            "std_format": "std_format",
+            "cxx_modules": "cxx_modules",
+            "string_view_ret": "static_constexpr_vars_in_constexpr_func",
+            "no_crtp": "explicit_this",
+        }
+
+    def _check_feature_supported(self, name, feature_name=name):
+        compiler = self.settings.compiler
+        cppstd = compiler.get_safe("cppstd", default_cppstd(self))
+        feature = self._feature_compatibility[feature_name]
+
+        # check C++ version
+        if not valid_min_cppstd(self, feature["std"]):
+            raise ConanInvalidConfiguration(
```

Comment:
I like how this is implemented, in conjunction with the following a few lines above
```cppstd = compiler.get_safe("cppstd", default_cppstd(self))```

Indeed, in Conan 1.x the `compiler.cppstd` setting is typically not added by default (although it is supported). In practice, this means that most Conan 1.x users will get a ConanInvalidConfiguration (because it seems that for the most part c++20 is required, and it is not the baseline/default version for any compiler yet as far as I know)

as a compromise, perhaps do the following:
- if `compiler.cppstd` is defined and lower than the minimum required, raise an exception
- if `compiler.cppstd` is NOT defined and `default_cppstd()` is lower than the minimum, perhaps just display a warning? The package can still be created (since it's header only) - however this delays the "error" until later. Do the `mp-units` headers evaluate `__cplusplus` or something similar to check for and warn when C++20 or higher is not enabled?

Reply:
Agree